packages/driver/src/cypress/command_queue.ts

```diff
@@ -557,4 +559,27 @@ export class CommandQueue extends Queue<$Command> {
 
     return promise
   }
+
+  // This function iterates through all upcoming commands in the queue, then
+  // discards the subject chain for every chainer that can't be referenced
+  // in the future (eg, no upcoming commands belong to the same chain).
+
+  // This is safe because aliases (which might be referenced later) are stored
+  // separately, in state('aliases'), and any subjects that "flow upwards" (eg.
+  // the subject of a chain inside a .then() command) have already replaced
+  // the subject of their parent chainer by the time this is called.
+  cleanSubjects () {
```

Comment:
Is this safe for commands that execute callbacks?

Reply:
Yep. Consider a test like this.

```
cy
  .get('button') // 1, chainer-1
  .then(btn => { // 2, chainer-1
    cy
      .find('form') // 3, chainer-2
      .click() // 4, chainer-2
  })
  .dblclick() // 5, chainer-1
```

So it ends up playing out like this:

1. Outer test function executes, resulting in command queue `[get, then, dblclick]` - the .then() callback hasn't been called yet, so .find and .click aren't in the queue.
2. .get resolves
  a. cleanSubjects() leaves the chainer-1 subject intact (.then and .dblclick are in the same chainer).
3. .then() executes the callback.
  a. Its commands are inserted into the queue, leaving it looking like `[get, then, find, click, dblclick]`.
  b. A link is created: the subject for chainer-2 is linked with the subject for chainer-1 (because this is how .then() works).
  c. .then resolves, cleanSubjects again leaves chainer-1 intact.
4. .find() resolves.
  a. Because of the link, chainer-1's subject (the button) is overwritten with the result of .find() (the form).
  b. cleanSubjects leaves chainer-1 (.dblclick references it) and chainer-2 intact (.click() references it)
5. .click() resolves.
  a. Because of the link, chainer-1's subject is again overwritten.
  b. cleanSubject leaves chainer-1 intact, but discards chainer-2 (no future commands reference it, the only future command is .dblclick)
6. .dblclick() resolves.
  a. cleanSubject discards chainer-2 (there are no future commands remaining)

A touch convoluted, because "a flat command queue" isn't the right abstraction for Cypress anymore - it really needs to be a nested structure. But that's work for another year.

Reply:
what about `.within()`?

Reply:
The `within` subject is stored in two places: cy.state('withinSubject') and in the closure of each individual command that uses it. Within doesn't really interact with cy.state('subjects') at all, and it already has its own cleanup mechanism inside the .within implementation.

(these are all good questions, I'm double-checking my work as I write these answers to be sure)

Reply:
my other thought --  if this runs after each command, this might be 'heavy' for tests with several tests.  is it possible when we break the subject chain if this can be cleaned up there? Maybe that's leaking implementation?